### PR TITLE
Add the LFSC printer

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the LFSC printer for proof nodes. This PR has been split to not include support for DSL rewrite rules yet, or its main print method which will be added in a followup PR.

Further PRs will connect this printer as a possible output format for cvc5.

### CODE REVIEW

File: src/proof/lfsc/lfsc_printer.cpp [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+}
+
+void LfscPrinter::printProofInternal(
+    LfscPrintChannel* out,
+    const ProofNode* pn,
+    const LetBinding& lbind,
+    const std::map<const ProofNode*, size_t>& pletMap,
+    std::map<Node, size_t>& passumeMap)
+{
+  // the stack
```
Comment (yoni206) [inline src/proof/lfsc/lfsc_printer.cpp:103]: Just making sure, you introduced a utility for such traversal / visits not so long ago. Can't it be used here?

File: src/proof/lfsc/lfsc_printer.cpp [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+  PfRule r = pn->getRule();
+  const std::vector<Node>& args = pn->getArguments();
+  std::vector<Node> as;
+  for (const Node& a : args)
+  {
+    Node ac = d_tproc.convert(a);
+    Assert(!ac.isNull());
+    as.push_back(ac);
+  }
+  PExprStream pf(pargs, d_tt, d_ff);
```
Comment (yoni206) [inline src/proof/lfsc/lfsc_printer.cpp:265]: Can you please generally document this block? Just to clarify its purpose.

File: src/proof/lfsc/lfsc_printer.cpp [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+      switch (lr)
+      {
+        case LfscRule::SCOPE: pf << h << as[2] << cs[0]; break;
+        case LfscRule::NEG_SYMM: pf << h << h << cs[0]; break;
+        case LfscRule::CONG: pf << h << h << h << h << cs[0] << cs[1]; break;
+        case LfscRule::AND_INTRO1: pf << h << cs[0]; break;
+        case LfscRule::NOT_AND_REV: pf << h << h << cs[0]; break;
+        case LfscRule::PROCESS_SCOPE: pf << h << h << as[2] << cs[0]; break;
+        case LfscRule::AND_INTRO2: pf << h << h << cs[0] << cs[1]; break;
+        // do not pass type (cs[0]->getResult()[0].getType())
```
Comment (yoni206) [inline src/proof/lfsc/lfsc_printer.cpp:425]: Can this be made an `Assert`?

File: src/proof/lfsc/lfsc_printer.cpp (hunk 1 above)
Comment (ajreynol) [reply src/proof/lfsc/lfsc_printer.cpp:103]: This is a new kind of traversal, one where the next things to print can either be a proof or a term (or a sort).

File: src/proof/lfsc/lfsc_printer.cpp (hunk 3 above)
Comment (ajreynol) [reply src/proof/lfsc/lfsc_printer.cpp:425]: This is leftover from when i added subtypes to the LFSC signature, this comment is no longer necessary.

File: src/proof/lfsc/lfsc_printer.cpp (hunk 2 above)
Comment (ajreynol) [reply src/proof/lfsc/lfsc_printer.cpp]: Done

File: src/proof/lfsc/lfsc_printer.cpp [hunk 4]
```diff
@@ -0,0 +1,518 @@
...
+          isLambda = (lr == LfscRule::LAMBDA);
+        }
+        if (r == PfRule::ASSUME)
+        {
+          // an assumption, must have a name
+          passumeIt = passumeMap.find(cur->getResult());
+          Assert(passumeIt != passumeMap.end());
+          out->printAssumeId(passumeIt->second);
+        }
+        else if (isLambda)
```
Comment (4tXJ7f) [inline src/proof/lfsc/lfsc_printer.cpp:150]:
```suggestion
        if (isLambda)
```

**Review by 4tXJ7f — APPROVED** [review]
LGTM, just minor suggestions.

File: src/proof/lfsc/lfsc_printer.cpp [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+      if (pit == processingChildren.end())
+      {
+        bool isLambda = false;
+        if (r == PfRule::LFSC_RULE)
+        {
+          Assert(!cur->getArguments().empty());
+          LfscRule lr = getLfscRule(cur->getArguments()[0]);
+          isLambda = (lr == LfscRule::LAMBDA);
+        }
+        if (r == PfRule::ASSUME)
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/proof/lfsc/lfsc_printer.cpp:143]: This actually changes the behavior.  We want to fall through to the else of this.

**Pull request merged**